src/main/java/io/appium/java_client/remote/SupportsLocation.java

```diff
@@ -16,19 +16,70 @@
 
 package io.appium.java_client.remote;
 
+import io.appium.java_client.CommandExecutionHelper;
+import io.appium.java_client.ExecutesMethod;
+import io.appium.java_client.MobileCommand;
 import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebDriverException;
 import org.openqa.selenium.html5.Location;
 import org.openqa.selenium.html5.LocationContext;
 import org.openqa.selenium.remote.html5.RemoteLocationContext;
 
-public interface SupportsLocation extends WebDriver, LocationContext {
-    public RemoteLocationContext getLocationContext();
+import java.util.Map;
+import java.util.Optional;
 
+public interface SupportsLocation extends WebDriver, ExecutesMethod, LocationContext {
+
+    @Deprecated(forRemoval = true)
+    RemoteLocationContext getLocationContext();
+
+    /**
+     * Gets the physical location of the browser.
+     *
+     * @return A {@link Location} containing the location information. Returns null if the location is not available
+     * @deprecated This method and whole {@link LocationContext} interface are deprecated, use {@link #getLocation()}
+     *     instead.
+     */
+    @Deprecated(forRemoval = true)
     default Location location() {
         return getLocationContext().location();
     }
 
+    /**
+     * Gets the physical location.
+     *
+     * @return A {@link Location} containing the location information. Throws {@link WebDriverException} if the
+     *     location is not available.
+     */
+    default io.appium.java_client.Location getLocation() {
+        Map<String, Number> result = CommandExecutionHelper.execute(this, MobileCommand.GET_LOCATION);
+        return Optional.ofNullable(result).map(r ->
+                new io.appium.java_client.Location(
+                        r.get("latitude").doubleValue(),
+                        r.get("longitude").doubleValue(),
+                        Optional.ofNullable(r.get("altitude")).map(Number::doubleValue).orElse(null)
+                )
+        ).orElseThrow(() -> new WebDriverException("Cannot retrieve location"));
```

Comment:
there is no need for special handling here. If the returned response is invalid then a "standard" error would probably be more helpful